### PR TITLE
Fix functor-like macro expansion

### PR DESCRIPTION
**Pull request by bruntib**
When there is a functor-like macro which is passed as parameter to another
"function" macro then its parameters are not listed at the place of expansion:

```cpp
#define foo(x) int bar() { return x; }
#define hello(fvar) fvar(0)
hello(foo)

int main() { 1 / bar(); }
```

Expansion of hello(foo) asserted Clang, because it expected an l_paren token in
the 3rd line after "foo", since it is a function-like token.

### CODE REVIEW

**Comment by dkrupp** [comment]
Can one of the admins verify this patch?

**Comment by martong** [comment]
Jenkins: run tests

File: lib/StaticAnalyzer/Core/PlistDiagnostics.cpp [hunk 1]
```diff
@@ -941,8 +943,10 @@ static MacroNameAndArgs getMacroNameAndArgs(SourceLocation ExpanLoc,
     return { MacroName, MI, {} };
 
   RawLexer.LexFromRawLexer(TheTok);
-  assert(TheTok.is(tok::l_paren) &&
-         "The token after the macro's identifier token should be '('!");
+  // When this is a functor-like token which is passed as parameter to another
+  // "function" token then its arguments are listed at another place.
```
Comment (Szelethus) [inline lib/StaticAnalyzer/Core/PlistDiagnostics.cpp]: Could you please clarify a bit more? I never heard of the term "functor-like token" before. What does "another place" refer to? Do we really not know?

File: lib/StaticAnalyzer/Core/PlistDiagnostics.cpp [hunk 2]
```diff
@@ -879,8 +879,10 @@ static std::string getMacroNameAndPrintExpansion(
 
         getMacroNameAndPrintExpansion(Printer, ArgIt->getLocation(), PP,
                                       Info.Args, AlreadyProcessedTokens);
-        if (MI->getNumParams() != 0)
-          ArgIt = getMatchingRParen(++ArgIt, ArgEnd);
+        if ((++ArgIt)->is(tok::l_paren))
```
Comment (Szelethus) [inline lib/StaticAnalyzer/Core/PlistDiagnostics.cpp:890]: I would prefer some comments here too.

**Review by Szelethus — COMMENTED** [review]
The change is fine, but again, macro expansion is the definition of a mess, both conceptually and in implementation, we really-really have to counterweight that with excessive amount of comments.

**Pull request merged**